README
** Arbitrary order Exchange-Correlation functional library **

Copyright Ulf Ekstrom <uekstrom@gmail.com> and contributors 2009-2010.
See http://admol.org/xcfun for more information.

The main interface is in include/xcfun.h
(or fortran/xc_fun_module.f90 for Fortran bindings).

Copying:

The library is licensed under the LGPL license version 3, see
COPYING.LESSER for more information.


Configuration:

Check that XC_MAX_ORDER is defined to the highest order derivatives
you need (and not higher) in src/config.h.  Using a too large value
for XC_MAX_ORDER makes compilation slow and the generated code huge.


Building a debug/development version:

Edit the Makefile that matches your compiler
(Makefile.gcc or Makefile.intel or ...)
to set CXX (C++ compiler) and flags and run
$ make -f Makefile.gcc
(or using the corresponding Makefile)
This will create the library file lib/libxcfun.so.


Building an optimized version:

Edit the Makefile that matches your compiler
(Makefile.gcc or Makefile.intel or ...)
and add -DNDEBUG to the compiler flags. Add optimization
compiler options. Make sure your compiler performs inlining
(-O3 with gcc).


CHANGELOG.md
# Change Log

## [Version 2.1.1] - 2020-11-12

### Changed

- Linux and macOS continuous integration testing is run on GitHub actions. See [PR #145](https://github.com/dftlibs/xcfun/pull/145)

### Fixed

- We polished the installation of header files, CMake target export files, and Python module. These are especially relevant for Conda packaging XCFun. See [PR #143](https://github.com/dftlibs/xcfun/pull/143)
- A numerical issue with SCAN functionals and small density gradients was fixed by James Furness (@JFurness1). See [issue #144](https://github.com/dftlibs/xcfun/issues/144) reported by Xing Zhang (@fishjojo) and subsequent [PR #146](https://github.com/dftlibs/xcfun/pull/146) for the fix.

## [Version 2.1.0] - 2020-09-18

- Many new functionals in the SCAN family have been added. Thanks to James
  Furness for the contribution.
  See [PR #140](https://github.com/dftlibs/xcfun/pull/140)
- The library is now available both as a Spack and a Conda package.
- The library can now be _natively_ compiled on Linux, macOS, and Windows.

### Changed

- **BREAKING** CMake >= 3.14 is required to configure the code.

## [Version 2.0.2] - 2020-07-15

### Fixed

- VWN3 functional has been fixed for the spin-polarized case. It previously gave wrong results
  when alpha and beta densities differed. Thanks to Zhenyu Zhu for reporting the problem
  and also suggesting the solution.
  See [PR #134](https://github.com/dftlibs/xcfun/pull/134) and 
  [issue #132](https://github.com/dftlibs/xcfun/issues/132).

## [Version 2.0.1] - 2020-05-06

### Fixed

- We removed the `DEBUG_POSTFIX` property from the properties on the `xcfun`
  target. This was leading to build failures when using the library through
  CMake `FetchContent` with mixed release/debug mode.

## [Version 2.0.0] - 2020-04-14

### Changed

- **BREAKING** The build system will only produce a shared (default) or static
  library. Compilation of the static library can be requested by setting
  `BUILD_SHARED_LIBS` to `OFF`.
- macOS CI testing was moved to Azure Pipelines.
- The dependency on pybind11 was bumped to v2.5.0

### Fixed

- We corrected a number of wrinkles in the handling of symbol visibility in the
  shared library.

## [Version 2.0.0a7] - 2020-04-10

### Fixed

- Address warnings from compilers. Fix #90.

## [Version 2.0.0a6] - 2020-02-23

### Fixed

- Compilation with GCC 5.4.0.

## [Version 2.0.0a5] - 2020-02-20

### Fixed

- Handling of 64-bit integers in the Fortran interface.

## [Version 2.0.0a4] - 2020-02-02

### Fixed

- The API function `xcfun_get` accepts a single in-out `double` parameter. It
  was erroneously declared to accept an array of `double`-s instead.

## [Version 2.0.0a3] - 2020-01-31

We have introduced a number of breaking changes, motivated by the need to
modernize the library. See the [migration guide](https://xcfun.readthedocs.io/en/latest/migration.html).

### Added

- Up-to-date API documentation generated with [Doxygen], [breathe], and [Sphinx].
- Up-to-date documentation on how to build and develop XCFun.
- Up-to-date documentation on how to use XCFun in your code.
- API functions `xcfun_which_vars` and `xcfun_which_mode`.
- A full example, based on CMake as build system generator, showing how to use
  the library from a C++ host. Thanks @stigrj!
- A full example, based on CMake as build system generator, showing how to use
  the library from a C host.
- A full example, based on CMake as build system generator, showing how to use
  the library from a Fortran host.

### Changed

- **BREAKING** All API functions are uniformly namespaced with the `xcfun_` prefix.
- **BREAKING** The Fortran interface has been completely rewritten using
  `iso_c_binding`: the library can now be compiled without the use of neither a
  C nor a Fortran compiler. :confetti_ball:
- **BREAKING** CMake option `XCFun_XC_MAX_ORDER` renamed to `XCFUN_MAX_ORDER`. New default value of 6.
- **BREAKING** CMake option `XCFun_ENABLE_PYTHON_INTERFACE` renamed to `XCFUN_PYTHON_INTERFACE`.

### Deprecated

### Removed

- **BREAKING** API functions `xc_serialize`, `xc_deserialize`, `xc_set_fromstring`, and `xc_derivative_index`.
- **BREAKING** The CMake options `ENABLE_FC_SUPPORT` and `ENABLE_64BIT_INTEGERS`.

### Fixed

### Security

## [Version 2.0.0a2] - 2020-01-21

## [Version 2.0.0a1] - 2019-12-15

### Added

- A user-friendly API function to set up functional evaluation `xc_user_eval_setup`. Thanks @ilfreddy.

### Changed

- **BREAKING** A compiler compliant with the C++11 (or later) standard is required.
- **BREAKING** CMake >= 3.11 is required to configure the code. 
- **BREAKING** The Python bindings are now generated using [pybind11] instead of
  SWIG. The dependency will be fetched at configuration time if not found on
  your system.
- **BREAKING** The Fortran interface is no longer build with the code, but
  shipped as a separate file to be compiled within your own Fortran code.

[Unreleased]: https://github.com/dftlibs/xcfun/compare/v2.1.1...HEAD
[Version 2.1.1]: https://github.com/dftlibs/xcfun/compare/v2.1.0...v2.1.1
[Version 2.1.0]: https://github.com/dftlibs/xcfun/compare/v2.0.2...v2.1.0
[Version 2.0.2]: https://github.com/dftlibs/xcfun/compare/v2.0.1...v2.0.2
[Version 2.0.1]: https://github.com/dftlibs/xcfun/compare/v2.0.0...v2.0.1
[Version 2.0.0]: https://github.com/dftlibs/xcfun/compare/v2.0.0a7...v2.0.0
[Version 2.0.0a7]: https://github.com/dftlibs/xcfun/compare/v2.0.0a6...v2.0.0a7
[Version 2.0.0a6]: https://github.com/dftlibs/xcfun/compare/v2.0.0a5...v2.0.0a6
[Version 2.0.0a5]: https://github.com/dftlibs/xcfun/compare/v2.0.0a4...v2.0.0a5
[Version 2.0.0a4]: https://github.com/dftlibs/xcfun/compare/v2.0.0a3...v2.0.0a4
[Version 2.0.0a3]: https://github.com/dftlibs/xcfun/compare/v2.0.0a2...v2.0.0a3
[Version 2.0.0a2]: https://github.com/dftlibs/xcfun/compare/v2.0.0a1...v2.0.0a2
[Version 2.0.0a1]: https://github.com/dftlibs/xcfun/releases/tag/v2.0.0a1

[GitHub]: https://github.com/dftlibs/xcfun

api/xcfun.F90
!
! XCFun, an arbitrary order exchange-correlation library
! Copyright (C) 2018 Ulf Ekström and contributors.
!
! This file is part of XCFun.
!
! This Source Code Form is subject to the terms of the Mozilla Public
! License, v. 2.0. If a copy of the MPL was not distributed with this
! file, You can obtain one at http://mozilla.org/MPL/2.0/.
!
! For information on the complete list of contributors to the
! XCFun library, see: <https://xcfun.readthedocs.io/>
!

module xcfun
  implicit none
!
  integer, parameter :: XCFUN_API_VERSION = 2

! Evaluation modes
  integer, parameter :: XC_PARTIAL_DERIVATIVES = 1
  integer, parameter :: XC_POTENTIAL = 2
  integer, parameter :: XC_CONTRACTED = 3

! Variable combinations
  integer, parameter :: XC_A = 0
  integer, parameter :: XC_N = 1
  integer, parameter :: XC_A_B = 2
  integer, parameter :: XC_N_S = 3
  integer, parameter :: XC_A_GAA = 4
  integer, parameter :: XC_N_GNN = 5
  integer, parameter :: XC_A_B_GAA_GAB_GBB = 6
  integer, parameter :: XC_N_S_GNN_GNS_GSS = 7
  integer, parameter :: XC_A_GAA_LAPA = 8
  integer, parameter :: XC_A_GAA_TAUA = 9
  integer, parameter :: XC_N_GNN_LAPN = 10
  integer, parameter :: XC_N_GNN_TAUN = 11
  integer, parameter :: XC_A_B_GAA_GAB_GBB_LAPA_LAPB = 12
  integer, parameter :: XC_A_B_GAA_GAB_GBB_TAUA_TAUB = 13
  integer, parameter :: XC_N_S_GNN_GNS_GSS_LAPN_LAPS = 14
  integer, parameter :: XC_N_S_GNN_GNS_GSS_TAUN_TAUS = 15
  integer, parameter :: XC_A_B_GAA_GAB_GBB_LAPA_LAPB_TAUA_TAUB = 16
  integer, parameter :: XC_A_B_GAA_GAB_GBB_LAPA_LAPB_TAUA_TAUB_JPAA_JPBB = 17
  integer, parameter :: XC_N_S_GNN_GNS_GSS_LAPN_LAPS_TAUN_TAUS = 18
  integer, parameter :: XC_A_AX_AY_AZ = 19
  integer, parameter :: XC_A_B_AX_AY_AZ_BX_BY_BZ = 20
  integer, parameter :: XC_N_NX_NY_NZ = 21
  integer, parameter :: XC_N_S_NX_NY_NZ_SX_SY_SZ = 22
  integer, parameter :: XC_A_AX_AY_AZ_TAUA = 23
  integer, parameter :: XC_A_B_AX_AY_AZ_BX_BY_BZ_TAUA_TAUB = 24
  integer, parameter :: XC_N_NX_NY_NZ_TAUN = 25
  integer, parameter :: XC_N_S_NX_NY_NZ_SX_SY_SZ_TAUN_TAUS = 26
  integer, parameter :: XC_A_2ND_TAYLOR = 27
  integer, parameter :: XC_A_B_2ND_TAYLOR = 28
  integer, parameter :: XC_N_2ND_TAYLOR = 29
  integer, parameter :: XC_N_S_2ND_TAYLOR = 30

! Indices into the output array of derivatives. Fortran numbering.
  integer, parameter :: XC_D0 = 1
  integer, parameter :: XC_D1 = 2
  integer, parameter :: XC_D2 = 3

! all lda derivatives up to order 4
  integer, parameter :: XC_D00 = 1
  integer, parameter :: XC_D10 = 2
  integer, parameter :: XC_D01 = 3
  integer, parameter :: XC_D20 = 4
  integer, parameter :: XC_D11 = 5
  integer, parameter :: XC_D02 = 6
  integer, parameter :: XC_D30 = 7
  integer, parameter :: XC_D21 = 8
  integer, parameter :: XC_D12 = 9
  integer, parameter :: XC_D03 = 10
  integer, parameter :: XC_D40 = 11
  integer, parameter :: XC_D31 = 12
  integer, parameter :: XC_D22 = 13
  integer, parameter :: XC_D13 = 14
  integer, parameter :: XC_D04 = 15

! gga up to order 3
! only derivatives that are nonzero at closed-shell reference
! with XC_VARS_NS are here
  integer, parameter :: XC_D00000 = 1
  integer, parameter :: XC_D10000 = 2
  integer, parameter :: XC_D00100 = 4
  integer, parameter :: XC_D00001 = 6
  integer, parameter :: XC_D20000 = 7
  integer, parameter :: XC_D10100 = 9
  integer, parameter :: XC_D10001 = 11
  integer, parameter :: XC_D02000 = 12
  integer, parameter :: XC_D01010 = 14
  integer, parameter :: XC_D00200 = 16
  integer, parameter :: XC_D00101 = 18
  integer, parameter :: XC_D00020 = 19
  integer, parameter :: XC_D30000 = 22
  integer, parameter :: XC_D20100 = 24
  integer, parameter :: XC_D12000 = 27
  integer, parameter :: XC_D11010 = 29
  integer, parameter :: XC_D10200 = 31
  integer, parameter :: XC_D10020 = 34
  integer, parameter :: XC_D02100 = 38
  integer, parameter :: XC_D01110 = 42
  integer, parameter :: XC_D00300 = 47
  integer, parameter :: XC_D00120 = 50

! tau mgga
! so far only linear response
! list under construction ...
  integer, parameter :: XC_D0000000 = 1
  integer, parameter :: XC_D1000000 = 2
  integer, parameter :: XC_D0010000 = 4
  integer, parameter :: XC_D0000100 = 6
  integer, parameter :: XC_D0000010 = 7
  integer, parameter :: XC_D0000001 = 8
  integer, parameter :: XC_D2000000 = 9
  integer, parameter :: XC_D1010000 = 11
  integer, parameter :: XC_D1000010 = 14
  integer, parameter :: XC_D1000001 = 15
  integer, parameter :: XC_D0200000 = 16
  integer, parameter :: XC_D0101000 = 18
  integer, parameter :: XC_D0100010 = 20
  integer, parameter :: XC_D0100001 = 21
  integer, parameter :: XC_D0020000 = 22
  integer, parameter :: XC_D0010010 = 25
  integer, parameter :: XC_D0010001 = 26
  integer, parameter :: XC_D0002000 = 27
  integer, parameter :: XC_D0001010 = 29
  integer, parameter :: XC_D0001001 = 30
  integer, parameter :: XC_D0000110 = 32
  integer, parameter :: XC_D0000101 = 33
  integer, parameter :: XC_D0000020 = 34
  integer, parameter :: XC_D0000011 = 35
  integer, parameter :: XC_D0000002 = 36

contains
  ! We pass strings as null terminated integer arrays to C, this
  ! should be portable if C and Fortran uses the same character set.
  subroutine str2ints(str, ints)
! ints must have length at least len(str)+1
    character, intent(in) :: str*(*)
    integer, intent(out) :: ints(*)
    integer i
    do i = 1, len_trim(str)
      ints(i) = ichar(str(i:i))
    enddo
    ints(len(str) + 1) = 0
  end subroutine str2ints

  subroutine ints2str(ints, str)
    character, intent(out) :: str*(*)
    integer, intent(in) :: ints(*)
    integer i, j
    i = 1
    do while (ints(i) .ne. 0)
      str(i:i) = char(ints(i))
      i = i + 1
    enddo
    do j = i, len(str)
      str(j:j) = ' '
    enddo
  end subroutine ints2str

  function xcfun_version()
    real(8) xcfun_version, xcfuve
    xcfun_version = xcfuve()
  end function xcfun_version

  subroutine xcfun_splash(text)
    character, intent(out) :: text*(*)
    integer :: ibuf(len(text) + 1)
    integer le
    le = len(text)
    call xcspla(ibuf, le)
    call ints2str(ibuf, text)
  end subroutine xcfun_splash

  function xcfun_sizeof_int()
    integer :: i, xcfun_sizeof_int
    if (huge(i) .eq. 2147483647) then
      xcfun_sizeof_int = 4
    else
      xcfun_sizeof_int = 8
    endif
  end function xcfun_sizeof_int

  ! Create a new, "empty" functional and return its id
  function xc_new_functional()
    integer :: xc_new_functional, xcnewf
    xc_new_functional = xcnewf(XCFUN_API_VERSION)
  end function xc_new_functional

  subroutine xc_free_functional(funid)
    integer funid
    call xcfree(funid)
  end subroutine xc_free_functional

  function xc_set(funid, param, val)
    character, intent(in) :: param*(*)
    integer, intent(in) :: funid
    integer :: xc_set, xcsets
    real(8), intent(in) :: val
    xc_set = xcsets(funid, val, len_trim(param), param)
  end function xc_set

  function xc_get(funid, param, val)
    character, intent(in) :: param*(*)
    integer, intent(in) :: funid
    integer :: xcgets, xc_get
    real(8), intent(out) :: val
    xc_get = xcgets(funid, val, len_trim(param), param)
  end function xc_get

  function xc_serialize(funid, buflen, buf)
    integer, intent(in) :: funid, buflen
    real(8), intent(out) :: buf(buflen)
    integer :: xc_serialize, xcseri
    xc_serialize = xcseri(funid, buflen, buf)
  end function xc_serialize

#ifdef XXX
  subroutine xc_short_description(param, description)
    integer, intent(in) :: param
    character, intent(out) :: description*(*)
    integer :: idescr(len(description) + 1)
    integer :: le
    le = len(description) + 1
    call xcssho(idescr, le, param)
    call ints2str(idescr, description)
  end subroutine

  subroutine xc_long_description(param, description)
    integer, intent(in) :: param
    character, intent(out) :: description*(*)
    integer :: idescr(len(description) + 1)
    integer :: le
    le = len(description) + 1
    call xcslon(idescr, le, param)
    call ints2str(idescr, description)
  end subroutine xc_long_description
#endif
  function xc_eval_setup(funid, vars, mode, order)
    integer :: funid, vars, mode, order, xc_eval_setup, xcevse
    xc_eval_setup = xcevse(funid, vars, mode, order)
  end function xc_eval_setup

  subroutine xc_eval(funid, npoints, densities, results)
    integer, intent(in) :: funid, npoints
    real(8), intent(in) :: densities(:, :)
    real(8), intent(out) :: results(:, :)

    integer :: i2

    i2 = 2
    IF (npoints .EQ. 1) i2 = 1

    call xceval(funid, npoints, densities(1, 1), densities(1, i2), &
                results(1, 1), results(1, i2))
  end subroutine xc_eval

  subroutine xc_eval_star(funid, npoints, m, densities, results)
    integer, intent(in) :: funid, npoints, m
    real(8), intent(in) :: densities(*)
    real(8), intent(out) :: results(:, :)
    call xceval(funid, npoints, densities(1), densities(m + 1), &
                results(1, 1), results(1, 2))
  end subroutine xc_eval_star

#if 0
  function xc_index(funid, exponents)
    integer, intent(in) :: exponents(*)
    integer funid, xc_index, xcdind
!radovan: it's easier to start in fortran with 1
!         it would be possible to start with 0
!         but then we would allocate 0:length-1
!         instead of 1:length
!         i find 0:length-1 allocations error-prone
!   xc_index = xcdind(funid,exponents)
    xc_index = xcdind(funid, exponents) + 1
  end function xc_index
#endif

  function xc_output_length(funid, order)
    integer, intent(in) :: funid, order
    integer xc_output_length, xcoule
    xc_output_length = xcoule(funid, order)
  end function xc_output_length
#if 0
  subroutine xc_param_name(setting_nr, name)
    integer, intent(in) :: setting_nr
    character, intent(out) :: name*(*)
    integer :: ibuf(len(name) + 1)
    integer :: le
    le = len(name) + 1
    call xcsnam(ibuf, le, setting_nr)
    call ints2str(ibuf, name)
  end subroutine xc_param_name
#endif
end module
